internal/vmpool/google/pool.go

```diff
@@ -0,0 +1,507 @@
+package google
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/drone-runners/drone-runner-aws/internal/vmpool"
+	"github.com/drone/runner-go/logger"
+
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/googleapi"
+)
+
+const (
+	provider = "google"
+)
+
+type googlePool struct {
+	init sync.Once
+
+	name        string
+	runnerName  string
+	credentials Credentials
+
+	os      string
+	rootDir string
+
+	// vm instance data
+	diskSize            int64
+	diskType            string
+	image               string
+	labels              map[string]string
+	network             string
+	subnetwork          string
+	project             string
+	privateIP           bool
+	scopes              []string
+	serviceAccountEmail string
+	size                string
+	tags                []string
+	zones               []string
+	userData            string
+	userDataKey         string
+
+	// pool size data
+	sizeMin int
+	sizeMax int
+
+	service *compute.Service
+}
+
+func (p *googlePool) GetName() string {
+	return p.name
+}
+
+func (p *googlePool) GetOS() string {
+	return p.os
+}
+
+func (p *googlePool) GetRootDir() string {
+	return p.rootDir
+}
+
+func (p *googlePool) GetMaxSize() int {
+	return p.sizeMax
+}
+
+func (p *googlePool) GetMinSize() int {
+	return p.sizeMin
+}
+
+func (p *googlePool) GetZone() string {
+	return p.zones[rand.Intn(len(p.zones))]
+}
+
+func (p *googlePool) Ping(ctx context.Context) error {
+	client := p.credentials.getService()
+	healthCheck := client.Regions.List(p.project).Context(ctx)
+	response, err := healthCheck.Do()
+	if err != nil {
+		return err
+	}
+	if response.ServerResponse.HTTPStatusCode == http.StatusOK {
+		return nil
+	}
+	return errors.New("unable to ping google")
+}
+
+func (p *googlePool) List(ctx context.Context) (busy, free []vmpool.Instance, err error) {
+	client := p.credentials.getService()
+
+	logr := logger.FromContext(ctx).
+		WithField("provider", provider).
+		WithField("pool", p.name)
+
+	list, err := client.Instances.List(p.project, p.GetZone()).Context(ctx).
+		Filter(fmt.Sprintf("labels.%s=%s", vmpool.TagCreator, p.runnerName)).
+		Filter(fmt.Sprintf("labels.%s=%s", vmpool.TagRunner, vmpool.RunnerName)).
+		Filter(fmt.Sprintf("labels.%s=%s", vmpool.TagPool, p.name)).
+		Do()
+
+	if list.Items == nil {
+		return
+	}
+
+	for _, vm := range list.Items {
+		logr.
+			WithField("machine", vm.Name).
+			WithField("id", vm.Id).
+			WithField("status", vm.Status).
+			Traceln("-------- list machines -----------")
```

Comment:
Maybe remove logging of every instance, every time the List method is called...?